core/ordersync/ordersync.go

```diff
@@ -0,0 +1,501 @@
+			if err != nil {
+				return err
+			}
+			rawReq = &rawRequest{
+				Type:         TypeRequest,
+				Subprotocols: []string{selectedSubprotocol.Name()},
+				Metadata:     encodedMetadata,
+			}
+		}
+
+		if err := json.NewEncoder(stream).Encode(rawReq); err != nil {
+			s.handlePeerScoreEvent(providerID, psUnexpectedDisconnect)
+			return err
+		}
+
+		rawRes, err := waitForResponse(ctx, stream)
+		if err != nil {
+			return err
+		}
+		s.handlePeerScoreEvent(providerID, psValidMessage)
+
+		subprotocol, found := s.subprotocols[rawRes.Subprotocol]
+		if !found {
+			s.handlePeerScoreEvent(providerID, psSubprotocolNegotiationFailed)
+			return fmt.Errorf("unsupported subprotocol: %s", subprotocol)
+		}
+		selectedSubprotocol = subprotocol
+		res, err := parseResponseWithSubprotocol(subprotocol, rawRes)
+		if err != nil {
+			s.handlePeerScoreEvent(providerID, psInvalidMessage)
+			return err
+		}
+
+		nextReq, err = subprotocol.HandleOrders(ctx, res)
+		if err != nil {
+			return err
+		}
+		s.handlePeerScoreEvent(providerID, receivedOrders)
+
+		if rawRes.Complete {
+			return nil
+		}
+	}
+}
+
+// shufflePeers randomizes the order of the given list of peers.
+func shufflePeers(peers []peer.ID) {
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(peers), func(i, j int) { peers[i], peers[j] = peers[j], peers[i] })
+}
+
+func waitForRequest(parentCtx context.Context, stream network.Stream) (*rawRequest, error) {
+	ctx, cancel := context.WithTimeout(parentCtx, requestResponseTimeout)
+	defer cancel()
+	reqChan := make(chan *rawRequest, 1)
+	go func() {
+		var rawReq rawRequest
+		if err := json.NewDecoder(stream).Decode(&rawReq); err != nil {
+			log.WithFields(log.Fields{
+				"error":     err.Error(),
+				"requester": stream.Conn().RemotePeer().Pretty(),
+			}).Warn("could not encode ordersync request")
+		}
+		reqChan <- &rawReq
+	}()
+
+	select {
+	case <-ctx.Done():
+		log.WithFields(log.Fields{
+			"error":     ctx.Err(),
+			"requester": stream.Conn().RemotePeer().Pretty(),
+		}).Warn("timed out waiting for ordersync request")
+		return nil, ctx.Err()
+	case rawReq := <-reqChan:
+		return rawReq, nil
+	}
+}
+
+func waitForResponse(parentCtx context.Context, stream network.Stream) (*rawResponse, error) {
+	ctx, cancel := context.WithTimeout(parentCtx, requestResponseTimeout)
+	defer cancel()
+	resChan := make(chan *rawResponse, 1)
+	go func() {
+		var rawRes rawResponse
+		if err := json.NewDecoder(stream).Decode(&rawRes); err != nil {
+			log.WithFields(log.Fields{
+				"error":    err.Error(),
+				"provider": stream.Conn().RemotePeer().Pretty(),
+			}).Warn("could not encode ordersync response")
+		}
+		resChan <- &rawRes
+	}()
+
+	select {
+	case <-ctx.Done():
+		log.WithFields(log.Fields{
+			"error":    ctx.Err(),
+			"provider": stream.Conn().RemotePeer().Pretty(),
+		}).Warn("timed out waiting for ordersync response")
+		return nil, ctx.Err()
+	case rawRes := <-resChan:
+		return rawRes, nil
+	}
+}
```

Comment:
Optional Suggestion: I think it's possible to de-dupe at least some of the code in these implementations. It would be easier with generics, but I think it could still be a bit cleaner.

Reply:
Can you offer a concrete suggestion? It's pretty hard to do without generics. We need to create a variable with a concrete type (either `rawResponse` or `rawResult`) in order for encoding/decoding to work correctly.

Reply:
We could put the log statements into functions. After looking again, that's the only change that can really be made, so it's totally up to you.